File: backend/App.Contracts/DTOs/UserDto.cs
using App.Domain.Entities;
using App.Domain.Enums;

namespace App.Contracts.DTOs;

public class UserDto
{
    // Parameterless constructor for deserialization
    public UserDto() { }
    
    public UserDto(UserEntity user, string bucketUrl)
    {
        Id = user.Id;
        Email = user.Email;
        UserTypeId = user.TypeId;
        UserType = user.Type?.TypeName;
        AccessLevel = user.Type?.AccessLevel;
        StudentCode = user.StudentCode;
        PhotoLink = user.PhotoPath != null ? bucketUrl + user.PhotoPath : null;
    }

    public Guid Id { get; set; }
    public string Email { get; set; } = string.Empty;
    public Guid UserTypeId { get; set; }
    public string? UserType { get; set; }
    public EAccessLevel? AccessLevel { get; set; }
    public string? StudentCode { get; set; }
    public string? PhotoLink { get; set; }
    
    public static List<UserDto> ToDtoList(List<UserEntity>? entities, string bucketUrl)
    {
        return entities?.Select(e => new UserDto(e, bucketUrl)).ToList() 
               ?? new List<UserDto>();
    }
}

File: mobile/lib/models/Responses/user_dto.dart
class UserDto {
  final String id;
  final String email;
  final String userTypeId;
  final String userType;
  final int accessLevel;
  final String studentCode;
  final String photoLink;

  UserDto({
    required this.id,
    required this.email,
    required this.userTypeId,
    required this.userType,
    required this.accessLevel,
    required this.studentCode,
    required this.photoLink,
  });

  factory UserDto.fromJson(Map<String, dynamic> json) {
    return UserDto(
      id: json['id'] ?? '',
      email: json['email'] ?? '',
      userTypeId: json['userTypeId'] ?? '',
      userType: json['userType'] ?? '',
      accessLevel: json['accessLevel'] ?? 0,
      studentCode: json['studentCode'] ?? '',
      photoLink: json['photoLink'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'email': email,
      'userTypeId': userTypeId,
      'userType': userType,
      'accessLevel': accessLevel,
      'studentCode': studentCode,
      'photoLink': photoLink,
    };
  }
}

